pkg/apis/rbac/v1/validation.go

```diff
@@ -0,0 +1,117 @@
+package rbacv1
+
+import (
+	"fmt"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+func (c *Config) Validate() error {
+	if err := c.validateRoles(); err != nil {
+		return err
+	}
+	if err := c.validateRoleBindings(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Config) validateRoles() error {
+	uniqueRoleIds := make(map[string]struct{})
+	for _, r := range c.GetRoles() {
+		roleId := r.GetId()
+		if roleId == "" {
+			return fmt.Errorf("role id cannot be empty")
+		}
+		if _, ok := uniqueRoleIds[roleId]; ok {
+			return fmt.Errorf("duplicate role id %q", roleId)
+		}
+		uniqueRoleIds[roleId] = struct{}{}
+		svcName := r.GetService()
+
+		// ensure service and method names exist
+		var svcDesc protoreflect.ServiceDescriptor
+		if maybeSvcDesc, err := protoregistry.GlobalFiles.FindDescriptorByName(protoreflect.FullName(r.GetService())); err == nil {
```

Comment:
Why are roles restricted to a service, again? Couldn't you just have the fullname in there (servicename.methodname) and fetch it with FindDescriptorByName as a protoreflect.MethodDescriptor?

Reply:
It's arbitrary, but it seemed like it would make the configuration more deliberate and less prone to accidental error or copy/paste mistakes.